src/lsptoolshost/buildDiagnosticsService.ts

```diff
@@ -0,0 +1,148 @@
+                currentFileBuildDiagnostics,
+                buildOnlyIds,
+                true
+            );
+            this._diagnosticsReportedByBuild.set(uri, buildDiagnostics);
+        }
+    }
+
+    public static filterDiagnosticsFromBuild(
+        diagnosticList: vscode.Diagnostic[],
+        buildOnlyIds: string[],
+        isDocumentOpen: boolean
+    ): vscode.Diagnostic[] {
+        const analyzerDiagnosticScope = languageServerOptions.analyzerDiagnosticScope as AnalysisSetting;
+        const compilerDiagnosticScope = languageServerOptions.compilerDiagnosticScope as AnalysisSetting;
+
+        // If compiler and analyzer diagnostics are set to "none", show everything reported by the build
+        if (analyzerDiagnosticScope === AnalysisSetting.None && compilerDiagnosticScope === AnalysisSetting.None) {
+            return diagnosticList;
+        }
+
+        // Filter the diagnostics reported by the build. Some may already be shown by live diagnostics.
+        const buildDiagnosticsToDisplay: vscode.Diagnostic[] = [];
+
+        // If it is a project system diagnostic (e.g. "Target framework out of support")
+        // then always show it. It cannot be reported by live.
+        const projectSystemDiagnostics = diagnosticList.filter((d) =>
+            BuildDiagnosticsService.isProjectSystemDiagnostic(d)
+        );
+        buildDiagnosticsToDisplay.push(...projectSystemDiagnostics);
+
+        // If it is a "build-only"diagnostics (i.e. it can only be found by building)
+        // then always show it. It cannot be reported by live.
+        const buildOnlyDiagnostics = diagnosticList.filter((d) =>
+            BuildDiagnosticsService.isBuildOnlyDiagnostic(buildOnlyIds, d)
+        );
+        buildDiagnosticsToDisplay.push(...buildOnlyDiagnostics);
+
+        // Check the analyzer diagnostic setting. If the setting is "none" or if the file is closed,
+        // then no live analyzers are being shown and bulid analyzers should be added.
+        // If FSA is on, then this is a no-op as FSA will report all analyzer diagnostics
+        if (
+            analyzerDiagnosticScope === AnalysisSetting.None ||
+            (analyzerDiagnosticScope === AnalysisSetting.OpenFiles && !isDocumentOpen)
+        ) {
+            const analyzerDiagnostics = diagnosticList.filter(
+                // Needs to be analyzer diagnostics and not already reported as "build only"
+                (d) => BuildDiagnosticsService.isAnalyzerDiagnostic(d) && !this.isBuildOnlyDiagnostic(buildOnlyIds, d)
+            );
+            buildDiagnosticsToDisplay.push(...analyzerDiagnostics);
+        }
+
+        // Check the compiler diagnostic setting. If the setting is "none" or if the file is closed,
+        // then no live compiler diagnostics are being shown and bulid compiler diagnostics should be added.
+        // If FSA is on, then this is a no-op as FSA will report all compiler diagnostics
```

Comment:
This approach sounds good to me as a starting point. I think we should dogfood the experience to confirm if the delay between build completion and background analysis reporting live diagnostics for open files (FSA off) or closed files (FSA on) is acceptable for real world solutions with large number of open files across the solution. Especially if there are compiler errors in the files which we are deferring to background analysis to report diagnostics. Otherwise, we may have to introduce some handshake between this diagnostic source and the existing live diagnostic source such that we report all build reported diagnostics for a file here until live background analysis has analyzed that file and reported live diagnostics for it.